src/Wit/Check.hs

```diff
@@ -2,40 +2,46 @@ module Wit.Check
   ( CheckError (..),
     parseFile,
     checkFile,
-    check0,
-    eitherIO,
+    check,
     Env,
     lookupEnv,
   )
 where
 
 import Control.Monad
+import Control.Monad.State
+import Control.Monad.Except
 import Data.Map.Lazy qualified as M
 import Data.Maybe
 import System.Directory
-import System.Exit (exitSuccess)
 import Text.Megaparsec
 import Wit.Ast
 import Wit.Parser (ParserError, pWitFile)
+import Prettyprinter
 
 data CheckError
-  = CheckError String (Maybe SourcePos)
-  | PErr ParserError
+  = PErr ParserError
+  | CheckError String (Maybe SourcePos)
+  | Bundle [CheckError]
 
-instance Show CheckError where
-  show (PErr bundle) = errorBundlePretty bundle
-  show (CheckError msg (Just pos)) = sourcePosPretty pos ++ ": " ++ msg
-  show (CheckError msg Nothing) = msg
+instance Pretty CheckError where
+  pretty (PErr bundle) = pretty (errorBundlePretty bundle)
+  pretty (CheckError msg (Just pos)) = pretty (sourcePosPretty pos) <> colon <+> pretty msg
+  pretty (CheckError msg Nothing) = pretty msg
+  pretty (Bundle es) = (vsep $ map pretty es) <> line
 
-type M = Either CheckError
+collect :: (MonadState [CheckError] m) => (MonadError CheckError m) => m () -> m ()
+collect ma = ma `catchError` (\e -> modify (\s -> e : s))
 
-report :: String -> M a
-report msg = Left $ CheckError msg Nothing
+report :: (MonadError CheckError m) => String -> m a
+report msg = throwError $ CheckError msg Nothing
 
-addPos :: SourcePos -> M a -> M a
-addPos pos ma = case ma of
-  Left (CheckError msg Nothing) -> Left (CheckError msg (Just pos))
-  ma' -> ma'
+addPos :: (MonadError CheckError m) => SourcePos -> m a -> m a
+addPos pos = withError updatePos
+  where
+    updatePos :: CheckError -> CheckError
+    updatePos (CheckError msg Nothing) = CheckError msg (Just pos)
+    updatePos e = e
```

Comment:
A not clear idea: maybe we need a quick rewrap function
```
bundleError :: (MonadState [CheckError] m) => (MonadError CheckError m) => m () -> m ()
bundleError m = do
  errs <- get
  throwError $ Bundle errs
```

and then we modify `addPos` to recursively handle `Bundle` case? What do you think @dm4?

The idea is about reduce repeat code template in `Main` and `CheckSpec`, I guess some experiment would helps.